Migrate OptionMenu and deps to graph-paper

openjck opened the pull request.

openjck commented on package.json:
```diff
@@ -32,6 +32,7 @@
     "jest-transform-svelte": "^2.1.1",
     "lerna": "^3.20.2",
     "prettier": "^2.0.2",
+    "prettier-plugin-svelte": "^0.7.0",
```
I switched from vscode to Neovim for now and this is needed to get Neovim's formatting plugin to correctly format Svelte files. I don't mind installing it globally if that's what we prefer, but we're going to need this here eventually and it may not be a bad idea to document the version that we're using.

openjck replied:
Again, happy to use global. That's fine. Just lmk.

openjck commented:
I still need to look into why the stories are not showing up. I'm probably missing something obvious.

openjck commented on packages/controls/package.json:
```diff
@@ -0,0 +1,19 @@
+  "scripts": {
+    "test": "echo \"Error: run tests from root\" && exit 1"
+  }
+}
```
I generated these by hand. Is there an easier way to do this with lerna?

hamilton replied:
Sort of! Though I find lerna in this regard to be cumbersome & doesn't do what I want. So take a look at `create.sh` :) That should guide you in the right direction.

openjck replied:
Excellent, thank you!

openjck commented on packages/icons/Caret.svelte:
```diff
@@ -0,0 +1,10 @@
+    fill="currentColor"
+    d="M8.59 16.59L13.17 12 8.59 7.41 10 6l6 6-6 6-1.41-1.41z" />
+  <path fill="none" d="M0 0h24v24H0V0z" />
+</svg>
```
Notice the new spelling. When we publish, we'll want to remember to use the word *caret* rather than *carat*.

hamilton replied:
Whoops! Good catch. Funny how I missed that.

openjck replied:
💎 

Not a big deal when we use the term internally, but we'll want to get the export right.

hamilton replied:
I'm reticent about putting the icons in this repository at any rate. Some of them are borrowed from material design, more from photon, and some I made myself. I was moving at a quick pace when these were added, so I lost track. We should be mindful of this from a licensing point of view.

hamilton replied:
`Caret.svelte` was adapted from Photon, it looks like, so the MIT license is good to go here. I think most of the icons are either Photon or I made them. Maybe we just add them as time goes on, and replace the material design ones which I don't think are MIT

openjck commented on packages/menu/FloatingMenu.svelte:
```diff
@@ -0,0 +1,81 @@
+  // FloatingMenu.svelte – handles the placement of menus when necessary.
+  // get the parent element location;
+  // set the position of this element based off the parent element.
+  // include offset.
```
Is this comment still accurate? In the repo it was copied from, the first line named the file as *Menu.svelte*, which isn't correct.

openjck commented:
Hamilton, I just wanted to double-check. Did you say that there was an issue where multiple `OptionMenu`s could be opened at the same timte, such that they would overlap each other. I put together a basic story with three `OptionMenu`s side by side and I actually can't reproduce the issue. I must be missing something.

openjck commented:
I don't think the *OptionMenu* component that I grabbed actually suffers from the issue where multiple menus can be open at the same time. Something like that does happen on GLAM, but that's a different component: *MenuList*

openjck commented:
Updated to fix stories and the multiple-menu issue. Ready for review.

hamilton commented on packages/controls/OptionMenu.svelte:
```diff
@@ -0,0 +1,270 @@
+  export let title;
+  export let options;
+  export let multi = false;
+  // current option must either be:
+  // (1) a javacsript object w/ a key / label, or
+  // (2) an array of objects shaped like (1).
```
```suggestion
```

hamilton commented on packages/controls/OptionMenu.svelte:
```diff
@@ -0,0 +1,270 @@
+  export let compact = false;
+  export let alignment = "center";
+  export let location = "bottom";
+  export let active = true; // is the button active?
```
```suggestion
  export let active = true;
```

hamilton commented on packages/controls/OptionMenu.svelte:
```diff
@@ -0,0 +1,270 @@
+  export let alignment = "center";
+  export let location = "bottom";
+  export let active = true; // is the button active?
+  export let menuActive; // is the menu currently active?
```
```suggestion
  export let menuActive;
```

hamilton commented on packages/controls/OptionMenu.svelte:
```diff
@@ -0,0 +1,270 @@
+    grid-auto-flow: column;
+    width: max-content;
+    grid-column-gap: var(--space-1h);
+    color: var(--subhead-gray-02);
```
hmm, let's replace this `--subhead-gray-02` color with the actual color. We don't tend to use these types of variables anymore and I think we ought to just go with the gray number.

hamilton commented on packages/controls/OptionMenu.svelte:
```diff
@@ -0,0 +1,270 @@
+    font-size: var(--text-015);
+    margin:0;
+    text-align: left;
+    /* min-width: var(--space-16x); */
```
```suggestion
```

hamilton commented on packages/controls/OptionMenu.svelte:
```diff
@@ -0,0 +1,270 @@
+    border-radius: var(--space-1h);
+  }
+
+  .down-carat {
```
let's make sure to rename these css classes to `caret`

hamilton commented on packages/icons/Caret.svelte:
```diff
@@ -0,0 +1,10 @@
+<svg
```
Over the weekend I added `ArrowheadDown` – could we use that icon in place of this one? They are the same but the namiong convention + attribution is clearer.

openjck replied:
Gotcha, so use *ArrowDown* instead of this icon? Do you want *Caret.svelte* in this repo at all?

hamilton commented on packages/icons/CheckBox.svelte:
```diff
@@ -0,0 +1,28 @@
+<script>
```
I'm fairly certain this & the empty check box icons come from https://material.io/resources/icons/?icon=check_box&style=baseline. Those are Apache licensed. I'm wondering if we should find an alternative or figure out what exactly to do here.

hamilton commented on packages/icons/DownCaret.svelte:
```diff
@@ -0,0 +1,18 @@
+<script>
```
We can use `ArrowheadDown` in the icons package here now.

hamilton commented on packages/menu/MenuList.svelte:
```diff
@@ -0,0 +1,70 @@
+<script>
+  // MenuList.svelte – handles the menu items specifically.
```
```suggestion
```

hamilton commented on packages/menu/MenuListItem.svelte:
```diff
@@ -0,0 +1,84 @@
+    user-select: none;
+  }
+
+  button {
```
these styles should be turned to classes and moved to `menu/style.css`

hamilton commented on packages/menu/package.json:
```diff
@@ -0,0 +1,19 @@
+  "name": "@graph-paper/menu",
+  "version": "0.0.0-alpha.5",
+  "description": "__DESCRIPTION_FORTHCOMING__",
+  "author": "hamilton <hamilton.ulmer@gmail.com>",
```
```suggestion
  "author": "Hamilton Ulmer <hamilton.ulmer@gmail.com>",
```

hamilton commented:
A few more notes from playing with this component in storybook:

- I wonder if the package here should be called `@graph-paper/optionmenu` and not `controls`.
- The multi menu should stay open after click on an option, I think? We should have a prop that is default to `true` for that. If that prop is `false`, it should close after clicking. This prop will be set to `false` for the single-item version.
- It’d be nice if we try to focus the stories toward being real-world useful. Let’s have the dropdown menus in the “Sticky w/ z-index” story have (1) color (single selector) and (1) style (multi-selector with bold, italic, uppercase, etc.) and have these affect the body style.
- The tooltip covers the menu when active. We can have the tooltip disappear when un-collapsed (and make that an option w/ default true)?
- I want `esc` to work on these menus, and they don’t seem to. Can we have esc collapse the menu when open?

openjck commented:
One other thing we talked about in our meeting earlier this week:

* Consider combining the "Show Title" and "Default" stories, making title a toggle

openjck commented on packages/controls/package.json:
```diff
@@ -0,0 +1,19 @@
+  "name": "@graph-paper/controls",
+  "version": "0.0.0-alpha.5",
+  "description": "__DESCRIPTION_FORTHCOMING__",
+  "author": "hamilton <hamilton.ulmer@gmail.com>",
```
Sure thing. Most of the other packages use *hamilton*, though. Should we fix those on master?

hamilton replied:
Yeah we should fix those as well

openjck replied:
Sweet. Gonna split that off into another issue.

https://github.com/graph-paper-org/graph-paper/issues/41

openjck commented:
I hope you don't mind, but I turned your comment into a checkbox for easier tracking

openjck commented:
Updated. I'll finish up the rest tomorrow.

openjck commented:
Updated and ready for re-review. Two notes:

1. Keeping the multi-menu open ended up being much more difficult than expected. I've split that off into #44.
2. I'll finish up the READMEs on Monday

hamilton commented on CHANGELOG.md:
```diff
@@ -2,6 +2,7 @@
 
 ## upcoming
 
+- Ports OptionMenu and its dependencies from moment-x-dashboard
```
```suggestion
- adds `@graph-paper/optionmenu` ([#19](https://github.com/graph-paper-org/graph-paper/pull/19/))
```
